File: backend/src/utils/catalog.mapper.js
const PROPERTY_KIND_LABELS = {
  homestay: 'Homestay',
  mini_hotel: 'Mini Hotel',
  villa: 'Villa',
  serviced_apartment: 'Serviced Apartment',
};

function num(value) {
  if (value == null) return 0;
  return Number(value);
}

function toPublicMapPin(pin) {
  if (!pin) return null;
  return {
    lat: num(pin.lat),
    lng: num(pin.lng),
    zoom: pin.zoom ?? 15,
    label: pin.label ?? null,
    pinBadge: pin.pinBadge ?? null,
    pinInfo: pin.pinInfo ?? null,
    googleMapsUrl: pin.googleMapsUrl,
    embedUrl: pin.embedUrl ?? null,
  };
}

function toPublicBranch(branch) {
  return {
    id: branch.code,
    dbId: branch.id,
    code: branch.code,
    propertyId: branch.propertyId,
    name: branch.name,
    address: branch.address,
    tagline: branch.tagline ?? null,
    priceFromVnd: num(branch.price),
    roomCount: branch.roomCount ?? 0,
    image: branch.imgUrl ?? null,
    isActive: branch.isActive,
    mapPin: toPublicMapPin(branch.mapPin),
    property: branch.property
      ? {
          id: branch.property.id,
          slug: branch.property.slug,
          name: branch.property.name,
          city: branch.property.city,
        }
      : undefined,
  };
}

function toPublicProperty(property, options = {}) {
  const { includeBranches = false, includeGallery = true, includeAmenities = true } = options;

  const base = {
    id: property.id,
    slug: property.slug,
    name: property.name,
    city: property.city,
    region: property.region,
    kind: property.kind,
    kindLabel: PROPERTY_KIND_LABELS[property.kind] ?? property.kind,
    tagline: property.tagline ?? null,
    description: property.description,
    address: property.address,
    priceFromVnd: property.priceFromVnd,
    roomCount: property.roomCount ?? 0,
    branchCount: property.branchCount ?? 0,
    rating: num(property.rating),
    reviewCount: property.reviewCount ?? 0,
    heroImageUrl: property.heroImageUrl ?? null,
    heroImage: property.heroImageUrl ?? null,
    highlights: Array.isArray(property.highlights) ? property.highlights : [],
    isActive: property.isActive,
    gallery: [],
    amenities: [],
    subBranches: [],
    reviews: [],
  };

  if (includeGallery && property.gallery?.length) {
    base.gallery = property.gallery.map((g) => g.imageUrl);
  }

  if (includeAmenities && property.amenities?.length) {
    base.amenities = property.amenities.map((row) => ({
      icon: row.amenity.icon,
      label: row.amenity.label,
    }));
  }

  if (includeBranches && property.branches?.length) {
    base.subBranches = property.branches.map(toPublicBranch);
  }

  return base;
}

const DEFAULT_ROOM_IMAGE =
  'https://images.unsplash.com/photo-1631049307264-da0ec9d70304?auto=format&fit=crop&w=1400&q=80';

const MAX_ROOM_GALLERY = 5;

function toDetailSlug(code) {
  return String(code).toLowerCase().replace(/[^a-z0-9]+/g, '-');
}

function parseStringList(raw) {
  if (!Array.isArray(raw)) return [];
  return raw.map((item) => String(item).trim()).filter(Boolean);
}

function buildRoomGallery(room, rt) {
  const out = [];
  const roomGallery = parseStringList(room.galleryImages);

  if (roomGallery.length) {
    out.push(...roomGallery.slice(0, MAX_ROOM_GALLERY));
  } else {
    if (room.imageUrl) out.push(room.imageUrl);
    if (rt?.gallery?.length) {
      for (const item of rt.gallery) {
        if (item.imageUrl && !out.includes(item.imageUrl) && out.length < MAX_ROOM_GALLERY) {
          out.push(item.imageUrl);
        }
      }
    }
  }

  if (!out.length) out.push(DEFAULT_ROOM_IMAGE);
  return out.slice(0, MAX_ROOM_GALLERY);
}

function toPublicRoom(room) {
  const rt = room.roomType;
  return {
    id: room.id,
    code: room.code,
    detailSlug: toDetailSlug(room.code),
    branchId: room.branch?.code ?? null,
    branchDbId: room.branchId,
    propertySlug: room.branch?.property?.slug ?? null,
    propertyId: room.branch?.propertyId ?? null,
    roomTypeId: room.roomTypeId,
    type: rt?.category ?? rt?.title ?? 'Room',
    roomTypeTitle: rt?.title ?? null,
    status: room.status,
    priceVnd: room.priceVnd,
    capacityLabel: rt?.capacityLabel ?? `Tối đa ${room.maxAdults} người lớn`,
    description: room.description,
    image: room.imageUrl ?? null,
    alt: room.altText ?? room.code,
    maxAdults: room.maxAdults,
    maxChildren: room.maxChildren,
    isActive: room.isActive,
  };
}

function toPublicRoomDetail(room) {
  const rt = room.roomType;
  const detailSlug = toDetailSlug(room.code);
  const branch = room.branch ? toPublicBranch(room.branch) : null;
  const property = room.branch?.property
    ? {
        id: room.branch.property.id,
        slug: room.branch.property.slug,
        name: room.branch.property.name,
        city: room.branch.property.city,
        region: room.branch.property.region ?? null,
        kind: room.branch.property.kind,
        kindLabel: PROPERTY_KIND_LABELS[room.branch.property.kind] ?? room.branch.property.kind,
        rating: num(room.branch.property.rating),
        reviewCount: room.branch.property.reviewCount ?? 0,
      }
    : null;

  const gallery = buildRoomGallery(room, rt);

  const extraParagraphs = parseStringList(room.extraParagraphs);
  const typeParagraphs = Array.isArray(rt?.paragraphs) ? rt.paragraphs.filter(Boolean) : [];
  const paragraphs = [];
  if (room.description) paragraphs.push(room.description);
  for (const p of extraParagraphs) {
    if (!paragraphs.includes(p)) paragraphs.push(p);
  }
  for (const p of typeParagraphs) {
    if (!paragraphs.includes(p)) paragraphs.push(p);
  }

  const amenities = rt?.amenities?.length
    ? rt.amenities.map((row) => ({
        icon: row.amenity.icon,
        label: row.amenity.label,
      }))
    : [];

  const policyBullets = Array.isArray(rt?.policyBullets)
    ? rt.policyBullets.filter(Boolean)
    : [];

  return {
    id: room.id,
    code: room.code,
    slug: detailSlug,
    detailSlug,
    badge: rt?.badge ?? 'Cherry House',
    title: rt?.title ?? room.code,
    subtitle: room.code,
    roomTypeTitle: rt?.title ?? null,
    roomTypeSlug: rt?.slug ?? null,
    category: rt?.category ?? null,
    areaSqm: rt?.areaSqm ?? 0,
    bedLabel: rt?.bedLabel ?? '',
    capacityLabel: rt?.capacityLabel ?? `Tối đa ${room.maxAdults} người lớn`,
    paragraphs,
    amenities,
    policyBullets,
    checkIn: rt?.checkInTime ?? '14:00',
    checkOut: rt?.checkOutTime ?? '12:00',
    priceVnd: room.priceVnd,
    basePriceVnd: rt?.basePriceVnd ?? room.priceVnd,
    status: room.status,
    maxAdults: room.maxAdults,
    maxChildren: room.maxChildren,
    gallery,
    image: gallery[0] ?? room.imageUrl ?? DEFAULT_ROOM_IMAGE,
    alt: room.altText ?? room.code,
    description: room.description,
    branch,
    property,
    propertySlug: property?.slug ?? null,
    branchCode: room.branch?.code ?? null,
    branchDbId: room.branchId,
    roomTypeId: room.roomTypeId,
    isActive: room.isActive,
  };
}

module.exports = {
  PROPERTY_KIND_LABELS,
  toPublicProperty,
  toPublicBranch,
  toPublicRoom,
  toPublicRoomDetail,
  toPublicMapPin,
  toDetailSlug,
  DEFAULT_ROOM_IMAGE,
  MAX_ROOM_GALLERY,
};

File: mobile/lib/data/catalog_mapper.dart
import 'package:flutter/material.dart';

import '../constants/catalog_constants.dart';
import '../models/models.dart';
import '../utils/media_url.dart';

PropertyKind _kindFromApi(String? raw) {
  switch (raw) {
    case 'mini_hotel':
      return PropertyKind.miniHotel;
    case 'villa':
      return PropertyKind.villa;
    case 'serviced_apartment':
      return PropertyKind.servicedApartment;
    default:
      return PropertyKind.homestay;
  }
}

IconData _amenityIcon(String? icon) {
  switch (icon) {
    case 'wifi':
      return Icons.wifi;
    case 'parking':
      return Icons.local_parking;
    case 'breakfast':
      return Icons.free_breakfast;
    case 'ac':
      return Icons.ac_unit;
    case 'pool':
      return Icons.pool;
    case 'kitchen':
      return Icons.kitchen;
    default:
      return Icons.check_circle_outline;
  }
}

RoomStatus _roomStatusFromOccupancy(String? occupancy) {
  switch (occupancy) {
    case 'held':
      return RoomStatus.pending;
    case 'booked':
      return RoomStatus.booked;
    case 'inactive':
      return RoomStatus.booked;
    default:
      return RoomStatus.available;
  }
}

RoomStatus _roomStatusFromInventory(String? status) {
  switch (status) {
    case 'pending':
      return RoomStatus.pending;
    case 'booked':
      return RoomStatus.booked;
    default:
      return RoomStatus.available;
  }
}

Property propertyFromApi(Map<String, dynamic> row) {
  final kind = _kindFromApi(row['kind'] as String?);
  final kindLabel = row['kindLabel'] as String? ?? CatalogConstants.kindOptions['homestay']!;
  final hero = resolveMediaUrl(
    (row['heroImageUrl'] ?? row['heroImage']) as String?,
  );

  final branches = (row['subBranches'] as List<dynamic>? ?? [])
      .whereType<Map<String, dynamic>>()
      .map(subBranchFromApi)
      .toList();

  final gallery = (row['gallery'] as List<dynamic>? ?? [])
      .map((e) => resolveMediaUrl(e?.toString()))
      .where((e) => e.isNotEmpty)
      .toList();

  final amenities = (row['amenities'] as List<dynamic>? ?? [])
      .whereType<Map<String, dynamic>>()
      .map(
        (a) => Amenity(
          icon: _amenityIcon(a['icon'] as String?),
          label: a['label'] as String? ?? '',
        ),
      )
      .where((a) => a.label.isNotEmpty)
      .toList();

  return Property(
    id: row['id'] as int?,
    slug: row['slug'] as String? ?? '',
    name: row['name'] as String? ?? '',
    city: row['city'] as String? ?? '',
    region: row['region'] as String? ?? '',
    kind: kind,
    kindLabel: kindLabel,
    tagline: row['tagline'] as String? ?? '',
    description: row['description'] as String? ?? '',
    address: row['address'] as String? ?? '',
    priceFromVnd: (row['priceFromVnd'] as num?)?.toInt() ?? 0,
    roomCount: (row['roomCount'] as num?)?.toInt() ?? 0,
    branchCount: (row['branchCount'] as num?)?.toInt() ?? branches.length,
    rating: (row['rating'] as num?)?.toDouble() ?? 0,
    reviewCount: (row['reviewCount'] as num?)?.toInt() ?? 0,
    heroImageUrl: hero.isNotEmpty ? hero : CatalogConstants.defaultHeroImage,
    gallery: gallery.isNotEmpty ? gallery : (hero.isNotEmpty ? [hero] : []),
    amenities: amenities,
    subBranches: branches,
    highlights: (row['highlights'] as List<dynamic>? ?? [])
        .map((e) => e.toString())
        .where((e) => e.isNotEmpty)
        .toList(),
  );
}

SubBranch subBranchFromApi(Map<String, dynamic> row) {
  final image = resolveMediaUrl(row['image'] as String?);
  return SubBranch(
    id: (row['id'] ?? row['code'] ?? '').toString(),
    dbId: row['dbId'] as int?,
    name: row['name'] as String? ?? '',
    address: row['address'] as String? ?? '',
    tagline: row['tagline'] as String? ?? '',
    roomCount: (row['roomCount'] as num?)?.toInt() ?? 0,
    priceFromVnd: (row['priceFromVnd'] as num?)?.toInt() ?? 0,
    imageUrl: image,
  );
}

Room roomFromApi(
  Map<String, dynamic> row, {
  String? occupancy,
  required String propertySlug,
  required String branchCode,
  int? branchDbId,
}) {
  final image = resolveMediaUrl(row['image'] as String?);
  final status = occupancy != null
      ? _roomStatusFromOccupancy(occupancy)
      : _roomStatusFromInventory(row['status'] as String?);

  return Room(
    id: '${row['id'] ?? row['code']}',
    code: row['code'] as String? ?? '',
    propertySlug: row['propertySlug'] as String? ?? propertySlug,
    branchId: row['branchId'] as String? ?? branchCode,
    detailSlug: row['detailSlug'] as String? ?? '',
    branchDbId: branchDbId ?? row['branchDbId'] as int?,
    type: row['type'] as String? ?? row['roomTypeTitle'] as String? ?? 'Room',
    status: status,
    priceVnd: (row['priceVnd'] as num?)?.toInt() ?? 0,
    capacityLabel: row['capacityLabel'] as String? ?? '',
    description: row['description'] as String? ?? '',
    imageUrl: image.isNotEmpty ? image : CatalogConstants.defaultHeroImage,
  );
}

List<Property> filterProperties(List<Property> list, BookingSearch search) {
  return list.where((p) {
    if (search.city.isNotEmpty && p.city != search.city) return false;
    if (search.kind != 'all') {
      final kindKey = switch (p.kind) {
        PropertyKind.homestay => 'homestay',
        PropertyKind.miniHotel => 'mini_hotel',
        PropertyKind.villa => 'villa',
        PropertyKind.servicedApartment => 'serviced_apartment',
      };
      if (kindKey != search.kind) return false;
    }
    return true;
  }).toList();
}

String isoDate(DateTime d) =>
    '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
